### PR TITLE
edge-case: handles reimports after updates

### LINKED ISSUE #49: [GraphSDK] [EdgeCase] Reimporting graph after having some updates inside UpdateTracker
## Steps to reproduce
1- Import the graph
2- Add new connections / Remove Connections
3- Reimport the graph again but this time the graph will already have those connections and basically have conflicts with Some of the updates inside **UpdateTracker**

## Acceptence Criterea
- On each graph import deconflict updates in tracker with graph
- Add tests

### PR DESCRIPTION
**Pull request by aramikm**
# Goal
The goal of this PR is to handle the edge case which re-imports of user data might happen after adding some updates. This would make sure that those updates are compatible with newly reimported user data and remove the non compatible ones.

Closes #49

# Notes
- Handles some of the unwraps around locks that were lingering around. We still have some unwraps in methods that do not return Result types but to handle them we will need to change the method signature which is  bigger than the scope of this PR.

# Checklist
- [x] Tests added

### CODE REVIEW

File: core/src/api/api.rs [hunk 1]
```diff
@@ -269,7 +275,7 @@ impl GraphAPI for GraphState {
 		Ok(self
 			.shared_state_manager
 			.read()
-			.unwrap()
+			.map_err(|_| DsnpGraphError::FailedtoReadLock(SHARED_STATE_MANAGER.to_string()))?
```
Comment (saraswatpuneet) [inline core/src/api/api.rs:278]: Do we want to also return error string or something here?

**Review by saraswatpuneet — APPROVED** [review]
lgtm! one comment

File: core/src/graph/updates.rs [hunk 2]
```diff
@@ -95,6 +95,28 @@ impl UpdateTracker {
...
+	) {
+		if let Some(arr) = self.updates.get(&schema_id) {
+			let mut synced_updates = arr.clone();
+			synced_updates.retain(|e| match e {
+				UpdateEvent::Add { dsnp_user_id, .. }
+					if existing_connections.contains(&dsnp_user_id) =>
+					false,
+				UpdateEvent::Remove { dsnp_user_id, .. }
+					if !existing_connections.contains(&dsnp_user_id) =>
+					false,
```
Comment (JoeCap08055) [inline core/src/graph/updates.rs:111]: 💯 Very cool... I wasn't aware of this type of match pattern

File: core/src/api/api.rs (hunk 1 above)
Comment (aramikm) [reply core/src/api/api.rs:278]: I'm not sure if I understand. Would you please explain more?
